### PR TITLE
o/ifacestate: move ResolveDisconnect to ifacestate

### PR DESCRIPTION
**Pull request by stolowski**
Move ResolveDisconnect from the repo up to ifacestate. The method is only used for resolving user input on snap disconnect and moving it outside makes it easy to re-use its higher-level logic for 'snap disconnect ---forget' with no need to duplicate its logic; only the two predicates declared locally: _connected_ and _connectedPlugOrSlot_ will need to be redefined to look up "conns" instead of repo (for context, respective followup PR is #8003).

### CODE REVIEW

File: overlord/ifacestate/ifacemgr.go [hunk 1]
```diff
@@ -288,6 +289,96 @@ func (m *InterfaceManager) ConnectionStates() (connStateByRef map[string]Connect
...
+			return false, err
+		}
+		return true, nil
+	}
+
+	connectedPlugOrSlot := func(snapName, plugOrSlotName string) ([]*interfaces.ConnRef, error) {
+		return m.repo.Connected(snapName, plugOrSlotName)
+	}
+
+	coreSnapName, _ := m.repo.GuessSystemSnapName()
```
Comment (stolowski) [inline overlord/ifacestate/ifacemgr.go]: I'll probably user mapper.SystemSnap here and keep GuessSystemName unexported.

File: interfaces/repo.go [hunk 2]
```diff
@@ -863,7 +811,7 @@ func (r *Repository) Connections(snapName string) ([]*ConnRef, error) {
 }
 
 // coreSnapName returns the name of the core snap if one exists
-func (r *Repository) guessSystemSnapName() (string, error) {
+func (r *Repository) GuessSystemSnapName() (string, error) {
```
Comment (pedronis) [inline interfaces/repo.go]: yes, let's try to avoid this

File: interfaces/repo.go [hunk 3]
```diff
@@ -283,19 +283,24 @@ func (r *Repository) Connection(connRef *ConnRef) (*Connection, error) {
 	// Ensure that such plug exists
 	plug := r.plugs[connRef.PlugRef.Snap][connRef.PlugRef.Name]
 	if plug == nil {
-		return nil, fmt.Errorf("snap %q has no plug named %q", connRef.PlugRef.Snap, connRef.PlugRef.Name)
+		return nil, &NoPlugOrSlotError{
```
Comment (pedronis) [inline interfaces/repo.go]: these *Error wrapping should have corresponding unit test changes, no?

File: interfaces/repo.go [hunk 4]
```diff
@@ -283,19 +283,24 @@ func (r *Repository) Connection(connRef *ConnRef) (*Connection, error) {
...
+			message: fmt.Sprintf("snap %q has no plug named %q",
+				connRef.PlugRef.Snap, connRef.PlugRef.Name)}
 	}
 	// Ensure that such slot exists
 	slot := r.slots[connRef.SlotRef.Snap][connRef.SlotRef.Name]
 	if slot == nil {
-		return nil, fmt.Errorf("snap %q has no slot named %q", connRef.SlotRef.Snap, connRef.SlotRef.Name)
+		return nil, &NoPlugOrSlotError{
```
Comment (pedronis) [inline interfaces/repo.go]: same,  also let's try to be consistent and adopt these errors in - afaict:

* connected
* ResolveConnect
* Disconnect

**Comment by zyga** [comment]
The basic change looks good. As for mapper I think it is a good idea but I wonder if it is better to go with Guess and iterate on subsequent changes (I would find that preferable). The same goes for the error return types (more smaller branches over one bigger).

File: interfaces/repo.go [hunk 5]
```diff
@@ -679,6 +607,26 @@ func (r *Repository) Connect(ref *ConnRef, plugStaticAttrs, plugDynamicAttrs, sl
...
+	message string
+}
+
+func (e *NotConnectedError) Error() string {
+	return e.message
+}
+
+// NoPlugOrSlotError is returned by Disconnect() if either the plug or slot does
+// no exist.
+type NoPlugOrSlotError struct {
```
Comment (zyga) [inline interfaces/repo.go]: This is currently unused beyond returning. Do you have plans for it?

File: interfaces/repo.go (hunk 3 above)
Comment (stolowski) [reply interfaces/repo.go]: Added

File: interfaces/repo.go (hunk 4 above)
Comment (stolowski) [reply interfaces/repo.go]: Done

File: interfaces/repo.go (hunk 5 above)
Comment (stolowski) [reply interfaces/repo.go]: Yes, in the followup. FWTW I've just discovered we had very similiar UnknownPlugOrSlotError before in this package, I've removed the old one in favor of this.

File: interfaces/repo.go (hunk 2 above)
Comment (stolowski) [reply interfaces/repo.go]: Done.

File: overlord/ifacestate/ifacemgr.go (hunk 1 above)
Comment (stolowski) [reply overlord/ifacestate/ifacemgr.go]: Done.

File: interfaces/repo.go [hunk 6]
```diff
@@ -862,7 +829,7 @@ func (r *Repository) Connections(snapName string) ([]*ConnRef, error) {
 	return conns, nil
 }
 
-// coreSnapName returns the name of the core snap if one exists
+// guessSystemSnapName returns the name of the system snap if one exists
```
Comment (zyga) [inline interfaces/repo.go]: Thanks for noticing this.

File: overlord/ifacestate/ifacestate_test.go [hunk 7]
```diff
@@ -176,10 +176,47 @@ type interfaceManagerSuite struct {
 	secBackend     *ifacetest.TestSecurityBackend
 	mockSnapCmd    *testutil.MockCmd
 	log            *bytes.Buffer
+	coreSnap       *snap.Info
```
Comment (zyga) [inline overlord/ifacestate/ifacestate_test.go]: Are there any plugs or slots to remove from `repo_test.go`? I didn't check, just asking.

Comment (stolowski) [reply overlord/ifacestate/ifacestate_test.go]: No, I don't think so, they are still used by remaining tests.

File: overlord/ifacestate/ifacemgr.go [hunk 8]
```diff
@@ -288,6 +289,96 @@ func (m *InterfaceManager) ConnectionStates() (connStateByRef map[string]Connect
...
+		}
+		return true, nil
+	}
+
+	connectedPlugOrSlot := func(snapName, plugOrSlotName string) ([]*interfaces.ConnRef, error) {
+		return m.repo.Connected(snapName, plugOrSlotName)
+	}
+
+	coreSnapName := SystemSnapName()
+	if coreSnapName == "" {
```
Comment (pedronis) [inline overlord/ifacestate/ifacemgr.go]: it could happen with guess but can this happen with SystemSnapName ?

Comment (zyga) [reply overlord/ifacestate/ifacemgr.go]: I don't think so. Although this made me realize we may have another problem. Repository would dynamically adjust to pick the right system snap name as snaps came and went. The code in the interface manager makes a one-time decision on snapd startup. I'm not sure this is correct with the new responsibilities. In the past if was only used for core configuration to handle things like "system" alias. Here it has more responsibility.

**Comment by stolowski** [comment]
I need to think about @zyga 's concern, marking blocked for now.

File: overlord/ifacestate/ifacemgr.go (hunk 8 above)
Comment (pedronis) [reply overlord/ifacestate/ifacemgr.go]: What happens in practice at the moment is that if we go from:

* no snapd to snapd
* or from core to snapd

there is a snapd restart involved,  so we do not meaningfully operate with the wrong value. This is true atm even for first boot installation of snapd.

Relevant code is:

https://github.com/snapcore/snapd/blob/master/overlord/snapstate/handlers.go#L1316
https://github.com/snapcore/snapd/blob/master/overlord/snapstate/handlers.go#L1351

Comment (stolowski) [reply overlord/ifacestate/ifacemgr.go]: > it could happen with guess but can this happen with SystemSnapName ?

Will address in a followup.

**Pull request merged**